### PR TITLE
[GOBBLIN-1706] Add DagActionStore to store the action to kill/resume one flow execution

### PR DESCRIPTION
**Pull request by ZihanLi58**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [ ] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1706

### Description
- [ ] Here are some details about my PR, including screenshots (if applicable):
Define DagActionStore
Add MysqlDagActionStore
Add new flow execution resource handle where we rely on dagactionstore to persist the action and use CDC stream to forward message

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
unit tests

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+      throws IOException {
+    try (Connection connection = this.dataSource.getConnection();
+        PreparedStatement insertStatement = connection.prepareStatement(String.format(INSERT_STATEMENT, tableName))) {
+      int i = 0;
+      insertStatement.setString(++i, flowGroup);
+      insertStatement.setString(++i, flowName);
+      insertStatement.setString(++i, flowExecutionId);
+      insertStatement.setString(++i, dagActionValue.toString());
+      insertStatement.executeUpdate();
+      connection.commit();
```
Comment (umustafi) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java:108]: why don't we also return a boolean for this function if the executeUpdate is successful? this should be consistent with the deleteDagAction right

**Review by umustafi — COMMENTED** [review]
generally looks good to me!

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java (hunk 1 above)
Comment (ZihanLi58) [reply gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java:108]: As long as add action does not throw exception, add should succeed. But delete can be true or false as even we try to delete some entry that does not exist, it does not throw exception, in this case, we need the boolean result to determine whether we delete the entry or not.

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  private final String tableName;
+  private static final String EXISTS_STATEMENT = "SELECT EXISTS(SELECT * FROM %s WHERE flow_group = ? AND flow_name =? AND flow_execution_id = ?)";
+
+  protected static final String INSERT_STATEMENT = "INSERT INTO %s (flow_group, flow_name, flow_execution_id, dag_action ) "
+      + "VALUES (?, ?, ?, ?)";
+  private static final String DELETE_STATEMENT = "DELETE FROM %s WHERE flow_group = ? AND flow_name =? AND flow_execution_id = ?";
+  private static final String GET_STATEMENT = "SELECT flow_group, flow_name, flow_execution_id, dag_action FROM %s WHERE flow_group = ? AND flow_name =? AND flow_execution_id = ?";
+  private static final String GET_ALL_STATEMENT = "SELECT flow_group, flow_name, flow_execution_id, dag_action FROM %s";
+  private static final String CREATE_TABLE_STATEMENT = "CREATE TABLE IF NOT EXISTS %s (" +
+  "flow_group varchar(128) NOT NULL, flow_name varchar(128) NOT NULL, flow_execution_id varchar(100) NOT NULL," +
```
Comment (Will-Lo) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java:51]: instead of hardcoding `128` length, can we use our global variables in case this number changes in the future?
I see references to `ServiceConfigKeys.MAX_FLOW_GROUP_LENGTH` e.g.
```
  private static final String SPECIFIC_CREATE_TABLE_STATEMENT = "CREATE TABLE IF NOT EXISTS %s (spec_uri VARCHAR("
      + FlowSpec.Utils.maxFlowSpecUriLength()
      + ") NOT NULL, flow_group VARCHAR(" + ServiceConfigKeys.MAX_FLOW_GROUP_LENGTH + "), flow_name VARCHAR("
      + ServiceConfigKeys.MAX_FLOW_GROUP_LENGTH + ")
```

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java [hunk 3]
```diff
@@ -296,7 +314,7 @@ synchronized public void stopDag(URI uri) throws IOException {
   private void killFlow(String flowGroup, String flowName, long flowExecutionId) throws IOException {
     int queueId =  DagManagerUtils.getDagQueueId(flowExecutionId, this.numThreads);
     String dagId = DagManagerUtils.generateDagId(flowGroup, flowName, flowExecutionId);
```
Comment (Will-Lo) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java:316]: If we have a dagId static class, it seems to make sense to me that we have a toString() functionality for this and we can remove `DagManagerUtils.generateDagId()`

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java [hunk 4]
```diff
@@ -369,6 +387,21 @@ public synchronized void setActive(boolean active) {
         for (Dag<JobExecutionPlan> dag : dags) {
           addDag(dag, false, false);
         }
+        if (dagActionStore.isPresent()) {
+          Collection<DagActionStore.DagAction> dagActions = dagActionStore.get().getDagActions();
```
Comment (Will-Lo) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java:391]: Do we need to worry about the scenario if we have multiple dagManagers or is that not in scope of this PR?

**Review by Will-Lo — CHANGES_REQUESTED** [review]

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java (hunk 4 above)
Comment (ZihanLi58) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java:391]: No, That will be future work, this PR is dedicated to make sure we can do warm standby and can forward kill/resume message

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+      throw new IOException("Please specify the config for MysqlDagActionStore");
+    }
+    this.tableName = ConfigUtils.getString(config, ConfigurationKeys.STATE_STORE_DB_TABLE_KEY,
+        ConfigurationKeys.DEFAULT_STATE_STORE_DB_TABLE);
+
+    this.dataSource = MysqlStateStore.newDataSource(config);
+    try (Connection connection = dataSource.getConnection();
+        PreparedStatement createStatement = connection.prepareStatement(String.format(CREATE_TABLE_STATEMENT, tableName))) {
+      createStatement.executeUpdate();
+      connection.commit();
```
Comment (arjun4084346) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java:70]: commit is unnecessary, every query is committed immediately unless `setAutoCommit` is set to `false`

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+  private final String tableName;
+  private static final String EXISTS_STATEMENT = "SELECT EXISTS(SELECT * FROM %s WHERE flow_group = ? AND flow_name =? AND flow_execution_id = ?)";
+
+  protected static final String INSERT_STATEMENT = "INSERT INTO %s (flow_group, flow_name, flow_execution_id, dag_action ) "
+      + "VALUES (?, ?, ?, ?)";
+  private static final String DELETE_STATEMENT = "DELETE FROM %s WHERE flow_group = ? AND flow_name =? AND flow_execution_id = ?";
+  private static final String GET_STATEMENT = "SELECT flow_group, flow_name, flow_execution_id, dag_action FROM %s WHERE flow_group = ? AND flow_name =? AND flow_execution_id = ?";
+  private static final String GET_ALL_STATEMENT = "SELECT flow_group, flow_name, flow_execution_id, dag_action FROM %s";
+  private static final String CREATE_TABLE_STATEMENT = "CREATE TABLE IF NOT EXISTS %s (" +
+  "flow_group varchar(" + ServiceConfigKeys.MAX_FLOW_GROUP_LENGTH + ") NOT NULL, flow_name varchar(" + ServiceConfigKeys.MAX_FLOW_GROUP_LENGTH + ") NOT NULL, flow_execution_id varchar(100) NOT NULL," +
```
Comment (arjun4084346) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java:52]: flow execution id also does not need 100 length

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+  protected final DataSource dataSource;
+  private final String tableName;
+  private static final String EXISTS_STATEMENT = "SELECT EXISTS(SELECT * FROM %s WHERE flow_group = ? AND flow_name =? AND flow_execution_id = ?)";
+
+  protected static final String INSERT_STATEMENT = "INSERT INTO %s (flow_group, flow_name, flow_execution_id, dag_action ) "
+      + "VALUES (?, ?, ?, ?)";
+  private static final String DELETE_STATEMENT = "DELETE FROM %s WHERE flow_group = ? AND flow_name =? AND flow_execution_id = ?";
+  private static final String GET_STATEMENT = "SELECT flow_group, flow_name, flow_execution_id, dag_action FROM %s WHERE flow_group = ? AND flow_name =? AND flow_execution_id = ?";
+  private static final String GET_ALL_STATEMENT = "SELECT flow_group, flow_name, flow_execution_id, dag_action FROM %s";
+  private static final String CREATE_TABLE_STATEMENT = "CREATE TABLE IF NOT EXISTS %s (" +
```
Comment (arjun4084346) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java:51]: What is the expectation if both KILL and RESUME are send for the same flow_name, flow_group, flow_execution_id?

Comment (arjun4084346) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java:51]: Will we setup retention on this table? What will be the basis of retention? Should we add a timestamp column or the flow_execution_id column is sufficient?

Comment (arjun4084346) [reply gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java:51]: I mean the caller can send KILL/RESUME when RESUME/KILL was already present, the insertion will fail because INSERT_STATEMENT does not have ON DUPLICATE KEY clause

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+  @Override
+  public boolean deleteDagAction(String flowGroup, String flowName, String flowExecutionId) throws IOException {
+    try (Connection connection = this.dataSource.getConnection();
+        PreparedStatement deleteStatement = connection.prepareStatement(String.format(DELETE_STATEMENT, tableName))) {
+      int i = 0;
+      deleteStatement.setString(++i, flowGroup);
+      deleteStatement.setString(++i, flowName);
+      deleteStatement.setString(++i, flowExecutionId);
+      int result = deleteStatement.executeUpdate();
+      connection.commit();
```
Comment (arjun4084346) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java:125]: same thing, L125 can be removed and return value can be returned inline

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/DagActionStore.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+  void addDagAction(String flowGroup, String flowName, String flowExecutionId, DagActionValue dagActionValue) throws IOException;
+
+  /**
+   * delete the dag action from {@link DagActionStore}
+   * @param flowGroup flow group for the dag action
+   * @param flowName flow name for the dag action
+   * @param flowExecutionId flow execution for the dag action
+   * @throws IOException
+   */
+  boolean deleteDagAction(String flowGroup, String flowName, String flowExecutionId) throws IOException;
```
Comment (arjun4084346) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/DagActionStore.java:75]: not an important one, but why deleteDagAction returns a boolean while addDagAction returns nothing?

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java (hunk 7 above)
Comment (ZihanLi58) [reply gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java:51]: We expect to fail the request. As I don't want two request overwrite with each other, so user need to wait until we finish last operation

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/DagActionStore.java (hunk 9 above)
Comment (ZihanLi58) [reply gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/DagActionStore.java:75]: As long as add action does not throw exception, add should succeed. But delete can be true or false as even we try to delete some entry that does not exist, it does not throw exception, in this case, we need the boolean result to determine whether we delete the entry or not.

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java (hunk 5 above)
Comment (ZihanLi58) [reply gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java:70]: yeah, I was thinking the same, but without this commit, unit test just fails and seems add operation does not take effect.

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java (hunk 6 above)
Comment (ZihanLi58) [reply gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java:52]: reduce to 13 which is same as dag state store

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/DagActionStore.java (hunk 9 above)
Comment (arjun4084346) [reply gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/DagActionStore.java:75]: Ah ok, let's update the javadoc with `@returns` then.

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java (hunk 6 above)
Comment (arjun4084346) [reply gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java:52]: looks like you put MAX_FLOW_EXECUTION_ID_LENGTH at incorrect place

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/restli/GobblinServiceFlowExecutionResourceHandlerWithWarmStandby.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+  @Override
+  public void resume(ComplexResourceKey<org.apache.gobblin.service.FlowStatusId, EmptyRecord> key) {
+    String flowGroup = key.getKey().getFlowGroup();
+    String flowName = key.getKey().getFlowName();
+    Long flowExecutionId = key.getKey().getFlowExecutionId();
+    try {
+      this.dagActionStore.addDagAction(flowGroup, flowName, flowExecutionId.toString(), DagActionStore.DagActionValue.RESUME);
+    } catch (IOException e) {
+    log.warn(String.format("Failed to add execution resume action for flow %s %s %s to dag action store due to", flowGroup, flowName, flowExecutionId), e);
+    throw new RestLiServiceException(HttpStatus.S_500_INTERNAL_SERVER_ERROR, e.getMessage());
```
Comment (arjun4084346) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/restli/GobblinServiceFlowExecutionResourceHandlerWithWarmStandby.java:58]: in case, kill is already present and user submitted this resume request, it will be considered user's mistake, in that case we should set error code to 4xx

**Review by arjun4084346 — APPROVED** [review]
some comments, lgtm otherwise

**Review by Will-Lo — APPROVED** [review]

**Review by umustafi — APPROVED** [review]

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java (hunk 5 above)
Comment (arjun4084346) [reply gobblin-runtime/src/main/java/org/apache/gobblin/runtime/dag_action_store/MysqlDagActionStore.java:71]: Oh, ok, it is because basicDataSource.setDefaultAutoCommit(false) in MysqlStateStore

**Review by arjun4084346 — APPROVED** [review]
